backend/drm/drm.c

```diff
@@ -0,0 +1,419 @@
+	wlr_log(L_INFO, "Scanning DRM connectors");
+
+	drmModeRes *res = drmModeGetResources(state->fd);
+	if (!res) {
+		wlr_log(L_ERROR, "Failed to get DRM resources");
+		return;
+	}
+
+	for (int i = 0; i < res->count_connectors; ++i) {
+		uint32_t id = res->connectors[i];
+
+		drmModeConnector *conn = drmModeGetConnector(state->fd, id);
+		if (!conn) {
+			wlr_log(L_ERROR, "Failed to get DRM connector");
+			continue;
+		}
+
+		struct wlr_output_state *output;
+		struct wlr_output *wlr_output;
+		int index = list_seq_find(state->outputs, find_id, &id);
+
+		if (index == -1) {
+			output = calloc(1, sizeof(struct wlr_output_state));
+			if (!state) {
+				drmModeFreeConnector(conn);
+				wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+				return;
+			}
+			wlr_output = output->wlr_output = wlr_output_create(&output_impl, output);
+			if (!wlr_output) {
+				drmModeFreeConnector(conn);
+				wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+				return;
+			}
+
+			output->renderer = &state->renderer;
+			output->state = DRM_OUTPUT_DISCONNECTED;
+			output->connector = id;
+			// TODO: Populate more wlr_output fields
+			// TODO: Move this to wlr_output->name
+			snprintf(output->name, sizeof(output->name), "%s-%"PRIu32,
+				 conn_name[conn->connector_type],
+				 conn->connector_type_id);
+
+			drmModeEncoder *curr_enc = drmModeGetEncoder(state->fd, conn->encoder_id);
+			if (curr_enc) {
+				output->old_crtc = drmModeGetCrtc(state->fd, curr_enc->crtc_id);
+				free(curr_enc);
+			}
+
+			list_add(state->outputs, output);
+			wlr_log(L_INFO, "Found display '%s'", output->name);
+		} else {
+			output = state->outputs->items[index];
+			wlr_output = output->wlr_output;
+		}
+
+		// TODO: move state into wlr_output
+		if (output->state == DRM_OUTPUT_DISCONNECTED &&
+			conn->connection == DRM_MODE_CONNECTED) {
+
+			wlr_log(L_INFO, "'%s' connected", output->name);
+			wlr_log(L_INFO, "Detected modes:");
+
+			for (int i = 0; i < conn->count_modes; ++i) {
+				struct wlr_output_mode_state *_state = calloc(1,
+						sizeof(struct wlr_output_mode_state));
+				_state->mode = conn->modes[i];
+				struct wlr_output_mode *mode = calloc(1,
+						sizeof(struct wlr_output_mode));
+				mode->width = _state->mode.hdisplay;
+				// TODO: Calculate more accurate refresh rate
+				// TODO: Check that this refresh rate is mHz
+				mode->height = _state->mode.vdisplay;
+				mode->state = _state;
+
+				wlr_log(L_INFO, "  %"PRIu16"@%"PRIu16"@%"PRIu32,
+					_state->mode.hdisplay, _state->mode.vdisplay,
+					_state->mode.vrefresh);
+
+				list_add(wlr_output->modes, mode);
+			}
+
+			output->state = DRM_OUTPUT_NEEDS_MODESET;
+			wlr_log(L_INFO, "Sending modesetting signal for '%s'", output->name);
+			wl_signal_emit(&state->backend->events.output_add, wlr_output);
+		} else if (output->state == DRM_OUTPUT_CONNECTED &&
+			conn->connection != DRM_MODE_CONNECTED) {
+
+			wlr_log(L_INFO, "'%s' disconnected", output->name);
+			// TODO: Destroy
```

Comment:
This was originally intended to just free some of the resources valid for an connected output (renderer and modes), and leave the rest of the structure in place, so we can reuse it if it gets connected again. If we're going to keep this design, we can leave the rest of the info there (name, id, etc.) and not have to free anything else.